### PR TITLE
[Docs] [Builtins] Add Note [Builtins and Plutus type checking]

### PR DESCRIPTION
**Pull request by effectfully**
I think this is the last piece of docs that I want to write for this module (I'll move some existing notes into this module, though).

### CODE REVIEW

File: plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs [hunk 1]
```diff
@@ -740,6 +737,97 @@ list [1] and that the type tag of the element being prepended equals the type ta
...
+            _                                                    -> EvaluationFailure
+
+but on the Plutus side its type is
+
+    integer -> integer
+
+because the @rep@ that 'SomeConstant' carries is 'Integer' in both the cases (in the type of the
+argument, as well as in the type of the result).
+
+This means that for this built-in function the Plutus type checker will accept a program that fails
```
Comment (zliu41) [inline plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:780]: > Plutus type checker will accept a program that fails at runtime

This is already the case for all builtin functions, right? Like many built-in functions expect constant arguments, and will fail at runtime if given a non-constant.

File: plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs [hunk 2]
```diff
@@ -99,6 +99,7 @@ data ExtensionFun
     | IdAssumeBool
     | IdAssumeCheckBool
     | IdSomeConstantBool
+    | IdIntegerAsBool
```
Comment (zliu41) [inline plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs:102]: Did you intend to add these? The PR title indicates that it is doc only.

File: plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs [hunk 3]
```diff
@@ -603,7 +603,7 @@ argument as a 'Bool', we could do that:
       where
         idAssumeCheckBoolPlc :: Opaque val Bool -> EvaluationResult Bool
         idAssumeCheckBoolPlc val =
-            case asConstant @_ @UnliftingError Nothing val of
+            case asConstant val of
```
Comment (zliu41) [inline plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:606]: Is it possible to have a builtin function whose plutus type is `bool -> bool`, which returns `EvaluationFailure` if the argument type is not `bool`, but also doesn't require the argument to be a constant? It seems none of `IdBool`, `IdAssumeBool` or `IdAssumeCheckBool` does that.

File: plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs (hunk 2 above)
Comment (effectfully) [reply plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs:102]: Yes, the title suggests that the PR is all about docs and we needed these examples in docs, hence they are added too.

File: plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs (hunk 3 above)
Comment (michaelpj) [reply plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:606]: I think that's impossible: the only *values* of type `bool` are constants.

File: plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs [hunk 4]
```diff
@@ -740,6 +737,97 @@ list [1] and that the type tag of the element being prepended equals the type ta
 the list [2] (extracted from the type tag for the whole list constant [1]).
 -}
 
+{- Note [Builtins and Plutus type checking]
+There's a direct correspondence between the Haskell type of the denotation of a builtin and the
+Plutus type of the builtin:
+
+1. elaboration turns a Haskell type variable into a concrete Haskell type representing a Plutus type
+   variable, which later becomes demoted to a regular Haskell value representing a Plutus type
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:745]: [Demotion](https://hackage.haskell.org/package/singletons-3.0.1/docs/Data-Singletons.html#v:demote) is the opposite of [promotion](https://downloads.haskell.org/~ghc/7.8.4/docs/html/users_guide/promotion.html). We use the same approach that the singletons library uses. A type representing a type variable is a type-level string for the name of the variable, a type-level number for the unique of the variable and the kind of variable. We demote all of that via `(KnownSymbol text, KnownNat uniq, KnownKind kind)`, of which the first two are from `base` (`singletons` uses them as well) and the last one is our own.

File: plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs (hunk 3 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:606]: > I think that's impossible: the only _values_ of type `bool` are constants.

The only `CekValue`s. We could have a lazy evaluator and pass `error` as an argument. The builtins machinery should be able to handle that and pass `error` through and the docs are about builtins.

**Review by michaelpj — APPROVED** [review]

**Review by bezirg — APPROVED** [review]

**Comment by effectfully** [comment]
Comments addressed, merging.

**Pull request merged**